game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/fire/FiringGroup.java

```diff
@@ -0,0 +1,143 @@
+  }
+
+  /**
+   * Splits up the firingUnits by suicideOnHit status and groups them by unit type
+   *
+   * @param name Name of the firing units
+   * @param firingUnits Collection of units that are firing
+   * @param targetUnits Collection of units that are being hit
+   * @return List of FiringGroup
+   */
+  public static List<FiringGroup> groupBySuicideOnHit(
+      final String name, final Collection<Unit> firingUnits, final Collection<Unit> targetUnits) {
+    final List<FiringGroup> groups = new ArrayList<>();
+    final List<Collection<Unit>> separatedBySuicide = separateSuicideOnHit(firingUnits);
+
+    // ensure each firing group has a unique name by adding prefixes
+    // if the firing groups have different types of suicide units
+    if (separatedBySuicide.size() == 1) {
+      groups.add(
+          new FiringGroup(
+              name,
+              name,
+              separatedBySuicide.get(0),
+              targetUnits,
+              Matches.unitIsSuicideOnHit().test(separatedBySuicide.get(0).iterator().next())));
+
+    } else if (separatedBySuicide.size() == 2
+        && Matches.unitIsSuicideOnHit().test(separatedBySuicide.get(0).iterator().next())
+            != Matches.unitIsSuicideOnHit().test(separatedBySuicide.get(1).iterator().next())) {
+      groups.addAll(
+          generateFiringGroupsWithOneSuicideAndOneNonSuicide(
+              name, targetUnits, separatedBySuicide));
+
+    } else {
+      groups.addAll(generateFiringGroups(name, targetUnits, separatedBySuicide));
+    }
+    return groups;
+  }
+
+  /**
+   * Separate the suicide on hit units from the others and group them by their type. The suicide on
+   * hit units need to fire separately so that they can be removed if they hit.
+   */
+  private static List<Collection<Unit>> separateSuicideOnHit(final Collection<Unit> units) {
+
+    final Map<UnitType, Collection<Unit>> map = new HashMap<>();
+    for (final Unit unit : CollectionUtils.getMatches(units, Matches.unitIsSuicideOnHit())) {
+      final UnitType type = unit.getType();
+      if (map.containsKey(type)) {
+        map.get(type).add(unit);
+      } else {
+        final Collection<Unit> unitList = new ArrayList<>();
+        unitList.add(unit);
+        map.put(type, unitList);
+      }
+    }
+
+    final List<Collection<Unit>> result = new ArrayList<>(map.values());
+    final Collection<Unit> remainingUnits =
+        CollectionUtils.getMatches(units, Matches.unitIsSuicideOnHit().negate());
+    if (!remainingUnits.isEmpty()) {
+      result.add(remainingUnits);
+    }
+    return result;
+  }
+
+  /** Handle the case where there are only two groups and one is suicide and the other is not */
+  private static List<FiringGroup> generateFiringGroupsWithOneSuicideAndOneNonSuicide(
+      final String name,
+      final Collection<Unit> targetUnits,
+      final List<Collection<Unit>> separatedBySuicide) {
+    final List<FiringGroup> groups = new ArrayList<>();
```

Comment:
WDYT of having a constructor do the work for the `isSuidideOnHit` check? 

In such a case this method would be reduced to:
```
    return separatedBySuicide.stream()
        .map(newFiringUnits -> new FiringGroup(name, newFiringUnits, targetUnits))
        .collect(Collectors.toList());
```

That makes it short enough it perhaps is worth inlining.

Reply:
You'll notice that the isSuicideOnHit is used to determine the suffix.  And just because it is a suicideOnHit doesn't mean it will have the same suffix.  If you look at the next method `generateFiringGroups`, it also checks the suicideOnHit but it uses a different suffix.

Here's some examples of firing unit groups and what their suffices would be:

One (doesn't matter if it is suicide or not) firing group -> No suffix
One suicide firing group and one non suicide firing group -> The suicide group has " suicide" suffix and the non suicide firing group has no suffix
Two suicide firing groups -> They will both have " suicide <unitType.name>" suffix.
Two suicide firing groups and one non suicide firing group -> The two suicide groups will have the suffix " suicide <unitType.name>" and the non suicide firing group will not have a suffix.